### PR TITLE
Freebsd fixes

### PR DESCRIPTION
**Pull request by infracaninophile**
Hi,

Thank you very much for open-sourcing citus.  When I saw the announcement, I just had to port it and add it to the official FreeBSD ports collection.  In doing so, I only found a small number of issues that affected compilation -- mostly to do with needing some additional header files and allowing the compiler to find postgresql's include files which are installed in a different place than for a typical Linux distribution.

There was one issue that would potentially cause compilation problems on a number of 32-bit platforms,
where this construct: INT64CONST(4294967296UL) could expand to 4294967296ULLL, which is a syntax error.

Once I've committed the port, within a few days it will become available as a official package available on all supported FreeBSD releases by:
{{{
  # pkg install pg_citus
}}}

### CODE REVIEW

**Comment by ozgune** [comment]
@infracaninophile, thanks for sending us a pull request for porting to FreeBSD. We're finalizing our process for taking contributions and testing. I'll chat with @sumedhpathak and @craigkerstiens on this -- we should be able to get to it in a few days.

**Comment by lfittl** [comment]
@infracaninophile First of all, apologies that we've been a bit slow regarding your patch here - it'll be automated in a few days, and will work similar to Facebook's process for external contributors on their repos.

Until then - in order to get your contribution in, I'd have to reach out to you for a quick digital signature by email, just so we have permission to use your code going forward.

Is it fine if I email you on the address in your git logs? (alternatively email me at lukas@citusdata.com)

We really appreciate your help & thanks in advance for understanding :)

**Comment by infracaninophile** [comment]
On 29/03/2016 22:04, Lukas Fittl wrote:
> @infracaninophile First of all, apologies that we've been a bit slow regarding your patch here - it'll be automated in a few days, and will work similar to Facebook's process for external contributors on their repos.
> 
> Until then - in order to get your contribution in, I'd have to reach out to you for a quick digital signature by email, just so we have permission to use your code going forward.
> 
> Is it fine if I email you on the address in your git logs? (alternatively email me at lukas@citusdata.com)
> 
> We really appreciate your help & thanks in advance for understanding :)
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/citusdata/citus/pull/407#issuecomment-203103762
> 

No problem.  I'd say that submitting a pull request pretty much implies
permission to use the code.  But if you want my explicit say-so, then:

Please do feel free to use the patches I sent in.

	Cheers,

	Matthew

**Comment by lfittl** [comment]
@infracaninophile Thank you for going through that - looking forward to make Citus work well on FreeBSD together :)

@ozgune Passing this to you / whomever wants to tackle the technical component here.

File: src/backend/distributed/worker/worker_partition_protocol.c [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #include "postgres.h"
 #include "funcapi.h"
 
+#include <netinet/in.h>
```
Comment (anarazel) [inline src/backend/distributed/worker/worker_partition_protocol.c]: Hm, aren't these in arpa/inet.h per posix? We should probably include both.

Comment (infracaninophile) [inline src/backend/distributed/worker/worker_partition_protocol.c]: On 2016/03/30 11:06, Andres Freund wrote:
>> @@ -17,6 +17,7 @@
>>  #include "postgres.h"
>>  #include "funcapi.h"
>>  
>> +#include <netinet/in.h>
> 
> Hm, aren't these in arpa/inet.h per posix? We should probably include both.

Either netinet/in.h or arpa/inet.h would work fine according to the man
page.  I chose netinet/in.h arbitrarily.

	Cheers,

	Matthew

File: configure.in [hunk 2]
```diff
@@ -58,6 +58,9 @@ SAVE_CFLAGS="$CFLAGS"
 AC_PROG_CC([$($PG_CONFIG --cc)])
 CFLAGS="$SAVE_CFLAGS"
 
+# Add the Postgres PKGINCLUDEDIR to the include search path
+CPPFLAGS="$CPPFLAGS -I$($PG_CONFIG --pkgincludedir)"
+
```
Comment (anarazel) [inline configure.in]: Two things:
a) for the benefit of odd platforms^WOSX we include the generated configure script in git (as does postgres).
b) I'm a bit confused about having to play with cppflags that way. postgres' Makefile.global.in, included via pgxs.mk, included via citus'  Makefile.global.in has:
```
pkgincludedir := $(shell $(PG_CONFIG) --pkgincludedir)
...
includedir_server = $(pkgincludedir)/server
includedir_internal = $(pkgincludedir)/internal
...
CPPFLAGS = @CPPFLAGS@

ifdef PGXS
override CPPFLAGS := -I$(includedir_server) -I$(includedir_internal) $(CPPFLAGS)
else # not PGXS
```
which files aren't found without the above?  I'd wager it's failures in connection_test.c? Per #411 I think the actual issue there is that internal/ has mistakenly been added to the include path.

File: src/backend/distributed/worker/worker_partition_protocol.c (hunk 1 above)
Comment (anarazel) [inline src/backend/distributed/worker/worker_partition_protocol.c]: On 2016-03-30 03:09:03 -0700, Matthew Seaman wrote:
> > Hm, aren't these in arpa/inet.h per posix? We should probably include both.
> 
> Either netinet/in.h or arpa/inet.h would work fine according to the man
> page.  I chose netinet/in.h arbitrarily.

IIRC it's actually different on different systems :(

File: configure.in (hunk 2 above)
Comment (infracaninophile) [inline configure.in]: On 2016/03/30 11:18, Andres Freund wrote:
>> @@ -58,6 +58,9 @@ SAVE_CFLAGS="$CFLAGS"
>>  AC_PROG_CC([$($PG_CONFIG --cc)])
>>  CFLAGS="$SAVE_CFLAGS"
>>  
>> +# Add the Postgres PKGINCLUDEDIR to the include search path
>> +CPPFLAGS="$CPPFLAGS -I$($PG_CONFIG --pkgincludedir)"
>> +
> 
> Two things:
> a) for the benefit of odd platforms^WOSX we include the generated configure script in git (as does postgres).
> b) I'm a bit confused about having to play with cppflags that way. postgres' Makefile.global.in, included via pgxs.mk, included via citus'  Makefile.global.in has:
> ```
> pkgincludedir := $(shell $(PG_CONFIG) --pkgincludedir)
> ...
> includedir_server = $(pkgincludedir)/server
> includedir_internal = $(pkgincludedir)/internal
> ...
> CPPFLAGS = @CPPFLAGS@
> 
> ifdef PGXS
> override CPPFLAGS := -I$(includedir_server) -I$(includedir_internal) $(CPPFLAGS)
> else # not PGXS
> ```
> which files aren't found without the above?  I'd wager it's failures in connection_test.c? Per #411 I think the actual issue there is that internal/ has mistakenly been added to the include path.

This is what was causing the compilation to fail:

% grep -r 'internal/c.h' .
./src/backend/distributed/test/connection_utils.c:#include "internal/c.h"

This seems to work -- at least, it compiles:

{{{
% git diff
diff --git a/configure.in b/configure.in
index 09aea8f..209c292 100644
--- a/configure.in
+++ b/configure.in
@@ -58,9 +58,6 @@ SAVE_CFLAGS="$CFLAGS"
 AC_PROG_CC([$($PG_CONFIG --cc)])
 CFLAGS="$SAVE_CFLAGS"

-# Add the Postgres PKGINCLUDEDIR to the include search path
-CPPFLAGS="$CPPFLAGS -I$($PG_CONFIG --pkgincludedir)"
-
 # check for a number of CFLAGS that make development easier

 # CITUSAC_PROG_CC_CFLAGS_OPT
diff --git a/src/backend/distributed/test/connection_utils.c
b/src/backend/distributed/test/connection_utils.c
index b5048cc..e944c1f 100644
--- a/src/backend/distributed/test/connection_utils.c
+++ b/src/backend/distributed/test/connection_utils.c
@@ -13,9 +13,9 @@
  *-------------------------------------------------------------------------
  */

-#include "internal/c.h"
+#include "c.h"
 #include "libpq-fe.h"
-#include "internal/libpq-int.h"
+#include "libpq-int.h"

 #include "distributed/test_helper_functions.h"

}}}

Let me push that to my branch and update the pull request.

	Cheers,

	Matthew

**Comment by ozgune** [comment]
I wanted to drop a quick note on ``src/backend/distributed/test/connection_utils.c``.

In #411, we're tracking build issues related to ``test/connection_utils.c``. @jasonmp85 is working on fixing this issue.

**Pull request closed without merging**